### ISSUE TITLE
struggling to find a way to differentiate between sessions in a group 

### ISSUE THREAD

**Issue by unphased** [tmux/tmux#1793]
### Issue description

Calling tmux from a 2nd session in a group results in the key-send landing in the first session.

Years ago, this actually worked as expected: The `tmux` commands would get sent to the most recently 'active' session. But since then the behavior of calling out to tmux e.g. from vim causes its determination of the target session to be more deterministically the first session now.

I believe that there is a way to force this to happen, and it's by passing various state all the way through the chain of programs that are involved... using an `#S` expansion, originating from a keybind, it WILL have the correct session. 

But it would be nice to be able to avoid having to do that. Consider the following: 

1. open terminal window 1. open tmux: `tmux`. Opens to session `0`
2. open terminal window 2. create the group: `tmux new -t 0`. Opens to session `0-1`; group `0` exists now, spawned out of session 0.
3. in terminal window 2 (in the shell) run `tmux display-message -p '#S'`. Witness the response of `0`.
4. in terminal window 2 type `:display-message -p '#S'`. Witness the response of `0-1`, as expected.
5. in terminal windwo 1, `:display-message -p '#S'` yields `0` as expected.

Step 3's response is indicative of the root problem, although it isn't exactly surprising or necessarily a bug... I just cannot think of why tmux shouldn't *always* default to targeting the most recently interacted-with session (as it used to in earlier versions of tmux).

**Comment by nicm** [comment]
This is odd because the activity time is definitely newer for the second session so it should pick that one:
~~~~
$ tmux ls -F '#{session_name} #{session_activity}'
0 1560327183
0-1 1560327225

**Comment by nicm** [comment]
Oh I bet it is $TMUX.

**Comment by unphased** [comment]
Yes so if I spawn a pane from 0-1 it will definitely begin to work 'as expected'. 

The issue is that the env var doesn't and can't in general get updated in realtime. Or can it? Should it?

**Comment by unphased** [comment]
ALSO. I should specify that I don't believe the intuition of targeting the most recently interacted-with session is the right one. Consider trying to run this from the shell: `sleep 5000 && tmux send-keys ...` From this situation the use of the environment variable to track which context it should use is sensible. It's just that I can't really tell when that environment should be getting clobbered... I used to have some zsh preexec code that would explicitly call `tmux show-env -s` but that strikes me as a total hack.

**Comment by nicm** [comment]
I'm not sure we should use $TMUX at all. If we don't use it when we know the pane from the tty but not when we don't, we end up with odd behaviour like this:
~~~~
$ tmux -Lx display -p '#S'
0-1
$ tmux -Lx display -p '#S' </dev/null
0

**Comment by nicm** [comment]
Using TMUX_PANE would be more sensible.

**Comment by nicm** [comment]
Try this please: [x.diff.txt](https://github.com/tmux/tmux/files/3280354/x.diff.txt)

~~~~
Index: cmd-find.c
===================================================================
RCS file: /cvs/src/usr.bin/tmux/cmd-find.c,v
retrieving revision 1.72
diff -u -p -r1.72 cmd-find.c
--- cmd-find.c	8 May 2019 18:05:03 -0000	1.72
+++ cmd-find.c	12 Jun 2019 08:29:22 -0000
@@ -76,38 +76,12 @@ static const char *cmd_find_pane_table[]
 	{ NULL, NULL }
 };
 
-/* Get session from TMUX if present. */
-static struct session *
-cmd_find_try_TMUX(struct client *c)
-{
-	struct environ_entry	*envent;
-	char			 tmp[256];
-	long long		 pid;
-	u_int			 session;
-	struct session		*s;
-
-	envent = environ_find(c->environ, "TMUX");
-	if (envent == NULL)
-		return (NULL);
-
-	if (sscanf(envent->value, "%255[^,],%lld,%d", tmp, &pid, &session) != 3)
-		return (NULL);
-	if (pid != getpid())
-		return (NULL);
-	log_debug("%s: client %p TMUX %s (session $%u)", __func__, c,
-	    envent->value, session);
-
-	s = session_find_by_id(session);
-	if (s != NULL)
-		log_debug("%s: session $%u still exists", __func__, s->id);
-	return (s);
-}
-
 /* Find pane containing client if any. */
 static struct window_pane *
 cmd_find_inside_pane(struct client *c)
 {
 	struct window_pane	*wp;
+	struct environ_entry	*envent;
 
 	if (c == NULL)
 		return (NULL);
@@ -116,6 +90,11 @@ cmd_find_inside_pane(struct client *c)
 		if (wp->fd != -1 && strcmp(wp->tty, c->ttyname) == 0)
 			break;
 	}
+	if (wp == NULL) {
+		envent = environ_find(c->environ, "TMUX_PANE");
+		if (envent != NULL)
+			wp = window_pane_find_by_id_str(envent->value);
+	}
 	if (wp != NULL)
 		log_debug("%s: got pane %%%u (%s)", __func__, wp->id, wp->tty);
 	return (wp);
@@ -880,8 +859,6 @@ cmd_find_from_mouse(struct cmd_find_stat
 int
 cmd_find_from_client(struct cmd_find_state *fs, struct client *c, int flags)
 {
-	struct session		*s;
-	struct winlink		*wl;
 	struct window_pane	*wp;
 
 	/* If no client, treat as from nothing. */
@@ -903,30 +880,6 @@ cmd_find_from_client(struct cmd_find_sta
 	if (wp == NULL)
 		goto unknown_pane;
 
-	/* If we have a session in TMUX, see if it has this pane. */
-	s = cmd_find_try_TMUX(c);
-	if (s != NULL) {
-		RB_FOREACH(wl, winlinks, &s->windows) {
-			if (window_has_pane(wl->window, wp))
-				break;
-		}
-		if (wl != NULL) {
-			log_debug("%s: session $%u has pane %%%u", __func__,
-			    s->id, wp->id);
-
-			fs->s = s;
-			fs->wl = s->curw; /* use current session */
-			fs->w = fs->wl->window;
-			fs->wp = fs->w->active; /* use active pane */
-
-			cmd_find_log_state(__func__, fs);
-			return (0);
-		} else {
-			log_debug("%s: session $%u does not have pane %%%u",
-			    __func__, s->id, wp->id);
-		}
-	}
-
 	/*
 	 * Don't have a session, or it doesn't have this pane. Try all
 	 * sessions.
@@ -948,17 +901,7 @@ cmd_find_from_client(struct cmd_find_sta
 	return (0);
 
 unknown_pane:
-	/*
-	 * We're not running in a known pane, but maybe this client has TMUX
-	 * in the environment. That'd give us a session.
-	 */
-	s = cmd_find_try_TMUX(c);
-	if (s != NULL) {
-		cmd_find_from_session(fs, s, flags);
-		return (0);
-	}
-
-	/* Otherwise we need to guess. */
+	/* We can't find the pane so need to guess. */
 	return (cmd_find_from_nothing(fs, flags));
 }
 
@@ -1006,6 +949,8 @@ cmd_find_target(struct cmd_find_state *f
 		strlcat(tmp, "CANFAIL,", sizeof tmp);
 	if (*tmp != '\0')
 		tmp[strlen(tmp) - 1] = '\0';
+	else
+		strlcat(tmp, "NONE", sizeof tmp);
 	log_debug("%s: target %s, type %s, item %p, flags %s", __func__,
 	    target == NULL ? "none" : target, s, item, tmp);

**Comment by unphased** [comment]
This does indeed work, and I am quite excited (due to the various integrations of mine that will magically begin working). But I am rather perplexed, because the value of $TMUX_PANE will be the same, in a given pane and yet now when I call `tmux display-message -p '#S'` in that same pane from different sessions the response shows it identifying the correct session.

**Comment by unphased** [comment]
I did the sleep test:

1. ran `sleep 5; tmux selectp -t:.+` from session 0-1
2. clicked into the terminal window of session 0 before 5 sec elapsed
3. observed the pane selection change occur in session 0

This indicates to me the timestamp is being used. 

I'm ok with that.

**Comment by nicm** [comment]
Changing to use TMUX_PANE doesn't affect how the session is worked out, it only affects how the pane is discovered if stdin isn't a tty.

If you run tmux from the shell it can try to use the ttyname or TMUX_PANE to find out the pane. If that works and it finds the pane, that pane can only be in one window, then it chooses the MRU session which contains that window.

If neither ttyname of TMUX_PANE works, tmux picks the MRU client and goes from there.

If you run a command from a key binding, or from the command prompt, tmux knows the client you were typing it.

A client can only be attached to one session, so it knows the session. Then that session can only have one current window and that window one active pane.

This all works OK for interactive use, but for scripting I recommend you explicitly specify the target you want with -t (that is why the -P flag to new/neww/splitw exists).

**Comment by unphased** [comment]
Awesome. That makes so much sense. Will this change make its way upstream? 😃

**Comment by nicm** [comment]
Yes I have applied it now, it will be in GitHub later on, as will the if-shell fix. Thanks!

**Comment by unphased** [comment]
Thank you!
